src/compiler/output-targets/test/custom-elements-types.spec.ts

```diff
@@ -0,0 +1,75 @@
+  config.namespace = 'TestApp';
+  config.buildEs5 = true;
+  config.globalScript = path.join(root, 'User', 'testing', 'src', 'global.ts');
+  config.outputTargets = [{ type: 'dist-custom-elements', dir: 'my-best-dir' }];
+
+  const bundleCustomElementsSpy = jest.spyOn(outputCustomElementsMod, 'bundleCustomElements');
+
+  compilerCtx.moduleMap.set('test', mockModule());
+
+  return { config, compilerCtx, buildCtx, bundleCustomElementsSpy };
+};
+
+describe('Custom Elements Typedef generation', () => {
+  it('should generate an index.d.ts file corresponding to the index.js file', async () => {
+    const componentOne = stubComponentCompilerMeta();
+    const componentTwo = stubComponentCompilerMeta({
+      componentClassName: 'MyBestComponent',
+      tagName: 'my-best-component',
+    });
+    const { config, compilerCtx, buildCtx } = setup();
+    buildCtx.components = [componentOne, componentTwo];
+
+    jest.spyOn(compilerCtx.fs, 'writeFile');
+
+    await generateCustomElementsTypes(config, compilerCtx, buildCtx, 'types_dir');
+
+    const expectedTypedefOutput = [
+      '/* TestApp custom elements */',
+      "export { StubCmp as StubCmp } from '../types_dir/components/stub-cmp/stub-cmp';",
+      "export { MyBestComponent as MyBestComponent } from '../types_dir/components/my-best-component/my-best-component';",
+      '',
+      '/**',
+      ' * Used to manually set the base path where assets can be found.',
+      ' * If the script is used as "module", it\'s recommended to use "import.meta.url",',
+      ' * such as "setAssetPath(import.meta.url)". Other options include',
+      ' * "setAssetPath(document.currentScript.src)", or using a bundler\'s replace plugin to',
+      ' * dynamically set the path at build time, such as "setAssetPath(process.env.ASSET_PATH)".',
+      ' * But do note that this configuration depends on how your script is bundled, or lack of',
+      ' * bundling, and where your assets can be loaded from. Additionally custom bundling',
+      ' * will have to ensure the static assets are copied to its build directory.',
+      ' */',
+      'export declare const setAssetPath: (path: string) => void;',
+      '',
+      'export interface SetPlatformOptions {',
+      '  raf?: (c: FrameRequestCallback) => number;',
+      '  ael?: (el: EventTarget, eventName: string, listener: EventListenerOrEventListenerObject, options: boolean | AddEventListenerOptions) => void;',
+      '  rel?: (el: EventTarget, eventName: string, listener: EventListenerOrEventListenerObject, options: boolean | AddEventListenerOptions) => void;',
+      '}',
+      'export declare const setPlatformOptions: (opts: SetPlatformOptions) => void;',
+      "export * from '../types_dir/components';",
+      '',
+    ].join('\n');
+
+    expect(compilerCtx.fs.writeFile).toBeCalledWith('my-best-dir/index.d.ts', expectedTypedefOutput, {
+      outputTargetType: 'dist-custom-elements',
+    });
+  });
```

Comment:
Can we clean up [our spy on `context.fs.writeFile`](https://github.com/ionic-team/stencil/pull/3400/files#diff-912e2efe974350fe77b7d3412b0fc0cbdaabbd62d4c9cdc99569d4ccd296d0baR40) here? Just in case we add additional tests to this file in the future?